wasm/validator/src/recursive/mod.rs

```diff
@@ -0,0 +1,286 @@
+/// 2. Validate every instruction in transaction, see next section.
+///
+/// # Instruction validation
+///
+/// Instruction validation is done with 3 steps:
+///
+/// 1. Check if instruction is some kind of [`If`](IfInstruction), [`Pair`] or etc.
+/// (see more in [`IndirectInstruction`](isi::IndirectInstruction)),
+/// if so then recursively unwrap it;
+/// 2. Check if there is a query inside instruction, if so then validate query
+/// with `validate_query` function;
+/// 3. Validate simple instruction (aka [`DirectInstruction`](isi::DirectInstruction)) with
+/// `validate_instruction` function;
+///
+/// # Query validation
+///
+/// Query validation is done directly with `validate_query` function.
+///
+/// # Example
+///
+/// ```
+///#![no_std]
+/// #![no_main]
+///
+/// #[cfg(not(test))]
+/// extern crate panic_halt;
+///
+/// use iroha_validator::prelude::*;
+///
+/// #[entrypoint(params = "[authority, operation]")]
+/// pub fn validate(
+///     authority: <Account as Identifiable>::Id,
+///     operation: NeedsPermissionBox,
+/// ) -> Verdict {
+///     validate_recursively(
+///         authority,
+///         operation,
+///         |_authority, _transaction| pass!(),
+///         validate_instruction,
+///         |_authority, _query| pass!(),
+///     )
+/// }
+///
+/// fn validate_instruction(
+///     authority: &<Account as Identifiable>::Id,
+///     instruction: DirectInstruction,
+/// ) -> Verdict {
+///     todo!("Your instruction validation logic here")
+/// }
+/// ```
+pub fn validate_recursively<T, I, Q>(
+    authority: <Account as Identifiable>::Id,
+    operation: NeedsPermissionBox,
+    validate_transaction: T,
+    validate_instruction: I,
+    validate_query: Q,
+) -> Verdict
+where
+    T: FnOnce(&<Account as Identifiable>::Id, tx::TransactionInfo) -> Verdict,
+    I: Fn(&<Account as Identifiable>::Id, isi::DirectInstruction) -> Verdict + Copy,
+    Q: Fn(&<Account as Identifiable>::Id, QueryBox) -> Verdict + Copy,
+{
+    match operation {
+        NeedsPermissionBox::Transaction(transaction) => {
+            let (transaction_info, instructions) = tx::TransactionInfo::disassemble_tx(transaction);
+            validate_transaction(&authority, transaction_info).and_then(|| {
+                let Executable::Instructions(instructions) = instructions else {
+                    pass!();
+                };
+
+                for isi in instructions {
+                    let verdict = validate_instruction_recursively(
+                        &authority,
+                        isi.clone(),
+                        validate_instruction,
+                        validate_query,
+                    );
+                    if verdict.is_deny() {
+                        return verdict;
+                    }
+                    isi.execute()
+                }
+                pass!();
+            })
+        }
+        NeedsPermissionBox::Instruction(instruction) => {
```

Comment:
when would we want to validate only one instruction?

Reply:
When user submits some WASM blob (e.g. smart-contract of trigger). We validate every instruction this WASM executes

Reply:
maybe we should have different host functions exposed for this. That way we'll open ourselves to the possibility for a wasm blob to directly call a function from the validator, without going through the host. We shouldn't mix these two

Reply:
> WASM will call function from another WASM? Interesting, never thought about this. However with that user WASM can just omit validation. I think we should still do that on host side

not exactly. Host is responsible for linking the two wasms. It will just link a wasm defined function instead of a host defined one.

Reply:
Sounds cool, will implement it then. But having multiple number of entrypoints sounds non-trivial

Reply:
> Sounds cool, will implement it then. But having multiple number of entrypoints sounds non-trivial
I didn't think you need another entrypoint, just another exported function

Reply:
> What about limits check? It either requires to still run this check on host side, or moving it to the validator

I think for the moment it's fine since it'll still be a host function. We can deal with that when we get to linking wasms

Reply:
> Sounds cool, will implement it then. But having multiple number of entrypoints sounds non-trivial

I don't think we necessarily need a different entrypoint if haven't needed it until now. The host will link all the functions based on what type of wasm it expects.

Reply:
> > What about limits check? It either requires to still run this check on host side, or moving it to the validator
> 
> I think for the moment it's fine since it'll still be a host function. We can deal with that when we get to linking wasms

actually, you can immediately link them without host or open an issue for that

Reply:
We decided to move it to #3371 . I close this discussion